README.md
# solarspeed Vue3 - Multipurpose Admin and Webapp Template

## ▶️ Getting started

First of all, Thank you so much for purchasing this template and for being our loyal customer. You are awesome! You are entitled to get free lifetime updates to this product and support from the cssninjaStudio team directly. **solarspeed** is a product built by [cssninjaStudio](https://cssninja.io/) and [Digisquad](https://digisquad.io/).

### Prerequisites

1. A recent web browser (Chrome, Edge, Firefox, ...)
2. [Nodejs LTS](https://nodejs.org/en/) _(> 14.x with npm >7)_ installed
3. Knowledge with [Typescript](https://github.com/microsoft/typescript) _(> 4.x)_ (should not be installed globally)
4. (recommended) [VSCode](https://code.visualstudio.com/ with [Volar](https://marketplace.visualstudio.com/items?itemName=johnsoncodehk.volar)

#### Install nodejs

First, check if you already have node and npm installed. To check if you have Node.js installed, run this command in your terminal:

```bash
node -v
```

To confirm that you have npm installed you can run this command in your terminal:

```bash
npm -v
```

If node is not installed on your machine, you can go to the official nodejs.org website, and choose the version depending on your operating system:

- <a href="https://nodejs.org/en/download/" target="_blank">Install node.js and npm on Windows, Linux or Mac OSX</a>

#### Check the npm version

npm is a separate project from Node.js, and tends to update more frequently. As a result, even if you’ve just downloaded Node.js (and therefore npm), you’ll probably need to update your npm. Luckily, npm knows how to update itself! To update your npm, type this into your terminal:

```bash
npm install --global npm@latest
```

#### Installing Yarn (recommended)

We personally recommend Yarn as we think it is a better package manager than npm. If you want to use Yarn, keep in mind that you should not mix the 2. You either use Yarn or Npm in a project, never both.

```bash
npm install --global yarn
```

## 📦 Dependencies installation

To setup the template and start installing project dependencies, run one of the following commands:

```bash
# using npm v7
npm install --legacy-peer-deps

# using yarn
yarn install
```

## 🔃 Run a development server

To start the development server, run one of the following commands:

```bash
# using npm v7
npm run dev

# using yarn
yarn dev
```

This will run the `dev` script from the `package.json` file.
You will notice that two servers are started: one for the frontend (vite) and one for the backend (json-server).

Vite is the build tool that we use to compile the frontend code.
It replace webpack and vue-cli, used in vue 2 ecosystem.
You can only start vite with `dev:vite`  
Read more about it [on vitejs.dev](https://vitejs.dev/)

Json-server is a fake REST-API server that we use to simulate the backend.
The configuration and the database are in the `/json-server` directory.
You can find how we use it in the `/src/pages/messaging-v1.vue` file and the `/src/composable/useChatApi.ts`  
You can only start vite with `dev:json-server`  
Read more about it [on github.com/typicode/json-server](https://github.com/typicode/json-server)

> - Access the solarspeed frontend in your browser at [http://localhost:3000/](http://localhost:3000/)
> - Access the Json-server backend in your browser at [http://localhost:8080/](http://localhost:8080/)

## 💡 What to do next ?

Our online documentation is a great place to learn how to use solarspeed.
We try to keep it mostly up to date, so you can always find the latest information.

> We also have a great [discord community](https://discord.cssninja.io/) where you can ask questions and show your work.

### [solarspeed on docs.cssninja.io](https://docs.cssninja.io/solarspeed?utm_source=readme)

- [Setup your project](https://docs.cssninja.io/solarspeed/documentation/setup-your-project.html?utm_source=readme)
- [Customizing solarspeed](https://docs.cssninja.io/solarspeed/documentation/customizing-solarspeed.html?utm_source=readme)
- [Going further](https://docs.cssninja.io/solarspeed/documentation/going-further.html?utm_source=readme)
- [Common Issues](https://docs.cssninja.io/solarspeed/documentation/common-issues.html?utm_source=readme)
- [Customer support](https://docs.cssninja.io/solarspeed/documentation/customer-support.html?utm_source=readme)
- [Changelog](https://docs.cssninja.io/solarspeed/documentation/changelog.html?utm_source=readme)

### [cssninjaStudio on youtube.com](https://www.youtube.com/channel/UCySf97KnXg9IvVeSeEvF-wQ)

- [Cleaning the project and getting started](https://www.youtube.com/watch?v=lE5VNpP8JqA)

We try to keep solarspeed dependencies and best practices up to dates. We also take care of user requests, either by implementing requested features, fixing bugs or improving the documentation.

If you want to contribute, feel free to create or up-vote discussions on the dedicated [public github repository](https://github.com/cssninjaStudio/solarspeed-public/discussions). You can also have access to the private github access, so you can view the source code history and submit issues. To do so, open a ticket on the [support portal](https://support.cssninja.io) with your github username.

### Useful links

- http://v3.vuejs.org/
  - https://v3.vuejs.org/guide/composition-api-introduction.html
  - https://v3.vuejs.org/api/composition-api.html
  - https://v3.vuejs.org/api/sfc-script-setup.html
  - https://v3.vuejs.org/guide/typescript-support.html
- https://vitejs.dev/
- https://pinia.esm.dev/
- https://vueuse.org/


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.2.0](https://github.com/cssninjaStudio/solarspeed/compare/v2.1.0...v2.2.0) (2022-01-03)

### Features

- **accessibility:** enable click-events-have-key-events eslint rule ([0d55212](https://github.com/cssninjaStudio/solarspeed/commit/0d55212e778e8268eeecc855a55ecdaf6ccb7353))
- **accessibility:** improve focus styles and keyboard navigation ([18a62e7](https://github.com/cssninjaStudio/solarspeed/commit/18a62e7cc973b75e2df49310ce9c88aea8d00846))
- **accessibility:** improve keyboard navigation on form pages ([3868a72](https://github.com/cssninjaStudio/solarspeed/commit/3868a72b72518dd7078362be6497a07c5321d5ab))
- add useThemeColors composable ([a1f84d8](https://github.com/cssninjaStudio/solarspeed/commit/a1f84d87d4c6c42ef1898726d9adc7c65249425a))
- **layouts:** add new Sideblock layout ([ff5f774](https://github.com/cssninjaStudio/solarspeed/commit/ff5f774b070fc6c09f74a712bdfbd03c75226627))
- **linters:** upgrade to stylelint v14 + other linter minor versions ([b5cbd8c](https://github.com/cssninjaStudio/solarspeed/commit/b5cbd8ce6627114c0a923fd49cc8c469ef10abe0))
- **page-icons:** add copy snippets to clipboard on all icons examples ([271acd9](https://github.com/cssninjaStudio/solarspeed/commit/271acd9406ea794252184aae8333fb6ff17bb3f0))
- **page-influencer:** split page into smaller and more reusable components ([46307a4](https://github.com/cssninjaStudio/solarspeed/commit/46307a4503ba352987dda72eb4c4511aad034bd8))
- **page-messaging-v1:** split page into smaller and more reusable components with api call samples ([0bf5d04](https://github.com/cssninjaStudio/solarspeed/commit/0bf5d04bec5cb0abf2a6ecacee123dc1590f01c5))
- **sidebar-collapsable-links:** replace has-nested-router-link directive with a component ([141311c](https://github.com/cssninjaStudio/solarspeed/commit/141311c00c73687b5fbac40e88e0fdd565a193d5))
- **stores:** convert activePanelState to pinia ([89c3c6a](https://github.com/cssninjaStudio/solarspeed/commit/89c3c6ac7166ec383e28f13fc4a02e497ddc90de))
- **stores:** convert activeSidebarState to pinia ([7a71043](https://github.com/cssninjaStudio/solarspeed/commit/7a7104343345afc1dfd45bc0db43ebe2416e3046))
- **stores:** convert darkModeState to pinia ([352b2ee](https://github.com/cssninjaStudio/solarspeed/commit/352b2eeca95357bd664a41dd611755095fb534f5))
- **stores:** convert navbarLayoutState and sidebarLayoutState to layoutSwitcher store ([949ed35](https://github.com/cssninjaStudio/solarspeed/commit/949ed357717d66ae66553d9c7e9aad172a5d516e))
- **stores:** enable hot module replacement on pinia modules ([37690dc](https://github.com/cssninjaStudio/solarspeed/commit/37690dcf26e3e5839015e78ccc79014b65fac671))
- **stores:** use viewWrapper store to handle pageTitle ([87e2766](https://github.com/cssninjaStudio/solarspeed/commit/87e2766f30bc954f767d0a9f1898227649946ee7))
- **styles:** Move styles to components to allow reusability ([#126](https://github.com/cssninjaStudio/solarspeed/issues/126)) ([448b53c](https://github.com/cssninjaStudio/solarspeed/commit/448b53ce69d97b94ac4054a6bf2b5e8d34f4d862))
- **v-flex-table-wrapper:** add pagination, sort, filter and async data to v-flex-table ([#127](https://github.com/cssninjaStudio/solarspeed/issues/127)) ([5a24ac2](https://github.com/cssninjaStudio/solarspeed/commit/5a24ac2b91216a7d29e34959bd54693a0bba27fe))
- **v-flex-table:** improve darkmode and responsive + add separators prop ([6aa8f5a](https://github.com/cssninjaStudio/solarspeed/commit/6aa8f5ad8049597096d95c8f75f9c892cfc68c8f))
- **v-flex-table:** rewrite component with slots and render functions ([cf67662](https://github.com/cssninjaStudio/solarspeed/commit/cf67662b0ce26a1618419a87a9d77ff8774986d3))
- **v-page-content-wrapper:** add base layout component ([47c90e2](https://github.com/cssninjaStudio/solarspeed/commit/47c90e27615eb1559ce8d3a294d21ac94968742c))
- **v-page-content:** add base layout component ([5881d07](https://github.com/cssninjaStudio/solarspeed/commit/5881d07e664c63f391030b404bce8fcd134dc622))
- **v-placeload:** add mobileWidth and mobileHeight properties ([17050a4](https://github.com/cssninjaStudio/solarspeed/commit/17050a4e586a113817010adc7227da378972813a))
- **v-view-wrapper:** add base layout component ([4c8384e](https://github.com/cssninjaStudio/solarspeed/commit/4c8384e5b4aa37c18ea14fe9b9516a8174fe8d6d))
- work on sideblock layout ([4aacc70](https://github.com/cssninjaStudio/solarspeed/commit/4aacc7010b74eeff670d20a54673ba8b7faff7ba))

### Bug Fixes

- add missing slots to base components ([9a64c83](https://github.com/cssninjaStudio/solarspeed/commit/9a64c83abe098499e4edfebdbe8535aef879fcb0))
- **composable:** export named composable instead of default ([a911ea4](https://github.com/cssninjaStudio/solarspeed/commit/a911ea4695a1427a297b0622fb53ac6a97693e1e))
- **composable:** update useApi error message ([9946588](https://github.com/cssninjaStudio/solarspeed/commit/994658809507a7746c37cf535badec706f3b5bb5))
- **composable:** wrap useNotif with createSharedComposable from vueuse ([d7a59cb](https://github.com/cssninjaStudio/solarspeed/commit/d7a59cbfcc5f91c07626acb68c23ada54ff0c137))
- **page-signup:** done button now submit the register form ([0ee5486](https://github.com/cssninjaStudio/solarspeed/commit/0ee5486d37630ac56ad421ef0d4693a1fddf630e)), closes [#124](https://github.com/cssninjaStudio/solarspeed/issues/124)
- sidebar hover labels ([1d21728](https://github.com/cssninjaStudio/solarspeed/commit/1d2172850b8ece264a3c0ecc113fe347ca1ba967))
- simplify app.ts + main.ts files and create solarspeedApp.vue ([e8b1132](https://github.com/cssninjaStudio/solarspeed/commit/e8b1132f2555a357b121a9253b1b43653f7cbae5))
- update modal form examples ([029dcd2](https://github.com/cssninjaStudio/solarspeed/commit/029dcd20c84309c2420a4a28c681784a482a7703))

vite.config.ts
import type { UserConfig, ConfigEnv } from 'vite'
import { loadEnv } from 'vite'
import path from 'path'
import Vue from '@vitejs/plugin-vue'
import Pages from 'vite-plugin-pages'
import Components from 'unplugin-vue-components/vite'
import ViteFonts from 'vite-plugin-fonts'
import ViteRadar from 'vite-plugin-radar'
import PurgeIcons from 'vite-plugin-purge-icons'
import { imagetools } from 'vite-imagetools'
import ImageMin from 'vite-plugin-imagemin'
import VueI18n from '@intlify/vite-plugin-vue-i18n'
import { VitePWA } from 'vite-plugin-pwa'
// import SolarSpeedDocumentation from './vite-plugin-solarspeed-doc/index'
// import envCompatible from 'vite-plugin-env-compatible'

export default ({ mode }: ConfigEnv): UserConfig => {
  const root = process.cwd()
  const env = loadEnv(mode, root)

  return {
    root: process.cwd(),
    mode: env.MODE,
    publicDir: 'public',
    logLevel: 'info',
    resolve: {
      alias: [
        // {
        //   find: '/~/',
        //   replacement: `/src/assets/`,
        // },
        {
          find: '/@src/',
          replacement: `/src/`,
        },
      ],
    },
    server: {
      host: '0.0.0.0',
      port: 3000,
      open: false,

      proxy: {
        '/api': {
          target: env.VITE_API_URL,
          secure: false,
          changeOrigin: true,
        },
        '/signupapi': {
          target: 'http://webback.thettdoc.com:3002',
          secure: false,
          changeOrigin: true,
          rewrite: (path) => path.replace(/^\/signupapi/, ''),
        },
        '/socket.io': {
          target: 'ws://webback.thettdoc.com:8080/socket.io',
          secure: false,
          ws: true,
        },
      },
    },

    build: {
      sourcemap: process.env.SOURCE_MAP === 'true',
      chunkSizeWarningLimit: 4000,
      rollupOptions: {
        external: [/\/demo\/.*/],
      },
    },
    css: {
      preprocessorOptions: {
        scss: {
          additionalData: ` 
            @import './src/scss/abstracts/_variables.scss';
            @import './src/scss/abstracts/_mixins.scss';
            @import './src/scss/pages/lists/_list-view-1.scss'; 
            @import './src/scss/pages/generic/_forms.scss'; 
            @import './src/scss/pages/generic/_utility.scss';
            @import './src/scss/pages/generic/_onboarding.scss';
            @import './src/scss/pages/generic/_widgets-lists.scss';
            @import './src/scss/layout/_navbar.scss';
            @import './src/scss/layout/_layout.scss';
            @import './src/scss/layout/_navbar-extension.scss';
            @import './src/scss/layout/_responsive.scss';
            @import './src/scss/layout/_sidebar-mobile.scss';
            @import './src/scss/layout/_sidebar-panel.scss';
            @import './src/scss/layout/_right-panel.scss';
            @import './src/scss/layout/_sidebar.scss'; `,
        },
      },
    },
    plugins: [
      // envCompatible(/* options */),
      Vue({
        include: [/\.vue$/],
      }),
      VueI18n({
        include: path.resolve(__dirname, './src/locales/**'),
      }),
      Pages({
        pagesDir: [
          {
            dir: 'src/pages',
            baseRoute: '',
          },
        ],
      }),
      // ViteComponents({
      //   extensions: ['vue', 'md'],
      //   dirs: ['documentation', 'src/components', 'src/layouts'],
      //   customLoaderMatcher: (path) => path.endsWith('.md'),
      // }),
      imagetools(),
      PurgeIcons(),
      Components({
        dirs: ['documentation', 'src/components', 'src/layouts'],
        extensions: ['vue', 'md'],
        dts: true,
        include: [/\.vue$/, /\.vue\?vue/, /\.md$/],
      }),
      ViteFonts({
        google: {
          families: [
            {
              name: 'Fira Code',
              styles: 'wght@400;600',
            },
            {
              name: 'Montserrat',
              styles: 'wght@500;600;700;800;900',
            },
            {
              name: 'Roboto',
              styles: 'wght@300;400;500;600;700',
            },
          ],
        },
      }),
      ViteRadar({
        analytics: {
          id: 'G-8PH6FM2JEL',
        },
      }),

      VitePWA({
        base: '/',
        includeAssets: [
          'favicon.svg',
          'favicon.ico',
          'robots.txt',
          'apple-touch-icon.png',
        ],
        manifest: {
          name: 'Solarspeed - A complete Vue 3 design system',
          short_name: 'Solarspeed',
          start_url: '/?utm_source=pwa',
          display: 'standalone',
          theme_color: '#ffffff',
          background_color: '#ffffff',
          icons: [
            {
              src: 'pwa-192x192.png',
              sizes: '192x192',
              type: 'image/png',
            },
            {
              src: 'pwa-512x512.png',
              sizes: '512x512',
              type: 'image/png',
            },
            {
              src: 'pwa-512x512.png',
              sizes: '512x512',
              type: 'image/png',
              purpose: 'any maskable',
            },
          ],
        },
      }),

      ImageMin({
        gifsicle: {
          optimizationLevel: 3,
          interlaced: false,
        },
        optipng: {
          optimizationLevel: 3,
        },
        mozjpeg: {
          quality: 90,
        },
        pngquant: {
          quality: [0.8, 1],
          speed: 4,
        },
        svgo: {
          plugins: [
            {
              name: 'removeViewBox',
              active: false,
            },
            {
              name: 'removeEmptyAttrs',
              active: false,
            },
          ],
        },
      }),
    ],
  }
}
